Newsletters: add 151 (2021-06-02) (#581)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2021-06-02-newsletter.md

---
title: 'Bitcoin Optech Newsletter #151'
permalink: /en/newsletters/2021/06/02/
name: 2021-06-02-newsletter
slug: 2021-06-02-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter describes a proposal to change Bitcoin Core's
transaction selection algorithm for miner block templates to slightly
increase miner profitability and give fee bumping users more
collective leverage.  Also included are our regular sections describing
software releases and release candidates, plus notable changes to popular
Bitcoin infrastructure software.

## News

- **Candidate Set Based (CSB) block template construction:** Mark
  Erhardt [posted][erhardt post] to the Bitcoin-Dev mailing list about
  an [analysis][es analysis] he and Clara Shikhelman performed on an
  alternative transaction selection algorithm for miners.  Bitcoin's
  consensus rules enforce that no transaction can be included in a block
  unless all of its unconfirmed ancestors are also included earlier in
  that same block.  Bitcoin Core addresses this constraint by treating
  each transaction with unconfirmed ancestors as if it contained both

>   the fees and the size of those ancestors.  For example, transaction B
>   depends on unconfirmed transaction A, so Bitcoin Core adds together
>   the fees paid by both transactions and divides them by the combined

Reviewer: Since you're introducing this hypothetical scenario here, then perhaps a conditional would be clearer:

```suggestion
  the fees and the size of those ancestors.  For example, if transaction B
  depends on unconfirmed transaction A, then Bitcoin Core would add together
  the fees paid by both transactions and divide them by the combined
```

    an aggregate effect on whether an ancestor transaction gets mined.

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- FIXME:harding hwi

- [LND 0.13.0-beta.rc3][LND 0.13.0-beta] is a release candidate that
  adds support for using a pruned Bitcoin full node, allows receiving
  and sending payments using Atomic MultiPath ([AMP][topic multipath payments]),
  and increases its [PSBT][topic psbt] capabilities, among other improvements
  and bug fixes.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1
repo], [Hardware Wallet Interface (HWI)][hwi repo],
[Rust Bitcoin][rust bitcoin repo], [BTCPay Server][btcpay server repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Bitcoin Core #20833][] is the first PR in an effort to bring [package

>   relay][topic package relay] to Bitcoin Core. This change allows the

Reviewer: I think it'd be better to emphasize that this PR is for mempool package acceptance. Future work may build package relay on top of this work, but best not to overpromise right now.

Reviewer 2: "Bitcoin Core #20833 is the first PR in an effort to implement mempool package acceptance in Bitcoin Core. This change allows the `testmempoolaccept` RPC to accept multiple transactions. Future PRs will enable testing L2 transaction chains, submitting transaction packages directly to the mempool through RPCs and communicating packages over the P2P network."

Reviewer 2: (emphasis on may)

Reviewer 3: Pushed a commit taking these suggestions and adding a contextual link for mempool package acceptance.

  `testmempoolaccept` RPC to accept multiple transactions. Future PRs will
  enable submitting transaction packages to the mempool through the RPC as well as
  specifications for communicating packages over the P2P network.

- [Bitcoin Core #22017][] Update Windows code signing certificate FIXME:jonatack

- [Bitcoin Core #18418][] wallet: Increase OUTPUT_GROUP_MAX_ENTRIES to 100 FIXME:Xekyo

> - [C-Lightning #4501][] adds [JSON schemas][] for the output of the
>   roughly half of C-Lightning's current commands (with schemas for the

Reviewer: ```suggestion
- [C-Lightning #4501][] adds [JSON schemas][] for the output of
  roughly half of C-Lightning's current commands (with schemas for the
```

  generate C-Lightning's documentation about what output each command
  produces.

- [LND #5025][] adds basic support for using [signet][topic signet].  Of
  the other LN implementations tracked by Optech, C-Lightning also has
  support for signet (see [Newsletter #117][news117 cl4068]).

- [LND #5155][] adds a configuration option to randomly select which
  wallet UTXOs to spend in a transaction; this reduces UTXO
  fragmentation in the wallet over time.  By contrast, the default coin
  selection algorithm in LND spends higher value UTXOs before lower
  value UTXOs; this minimizes fees in the short term but may result in
  needing to pay higher fees in the future when all inputs near the size
  of a transaction, or larger, have already been spent.

- [BOLTs #672][] BOLT 2: `option_shutdown_anysegwit` FIXME:dongcarl

- [BOLTs #872][] updates [BOLT3][] to specify in more detail the sort order
  to use for commitment transaction inputs and outputs.  This was
  originally specified by reference to [BIP69][], but one commentator
  points out that has caused three separate problems that lead to
  accidental channel closures and small amounts of funds lost to

>   unnecessary onchain fees.  The commetator suggests that this is

Reviewer: The "originally specified" (past tense) -> "points out"/"suggests" (present tense) feels a little awkward to me.

  consensus rules enforce that no transaction can be included in a block
  unless all of its unconfirmed ancestors are also included earlier in
  that same block.  Bitcoin Core addresses this constraint by treating
  each transaction with unconfirmed ancestors as if it contained both
  the fees and the size of those ancestors.  For example, if transaction B
  depends on unconfirmed transaction A, then Bitcoin Core adds together
  the fees paid by both transactions and divides them by the combined
  size of both transactions.  This allows Bitcoin Core to fairly compare
  all transactions in the mempool based on their effective feerate
  whether or not those transactions have any ancestors.

    However, Erhardt and Shikhelman note that a more sophisticated
    algorithm that may require a bit more CPU can find sets of related
    transactions that are even more profitable to mine than Bitcoin
    Core's existing simple algorithm.  The authors tested their
    algorithm on historic mempool data and found that it would've
    collected slightly more fees than Bitcoin Core's existing algorithm
    in almost all recent blocks.

>     If implemented and used by miners, the improved algorithm could

Author: That's my normal indenting practice, how is this the first time you noticed it?  :-D

  produces.

- [LND #5025][] adds basic support for using [signet][topic signet].  Of
  the other LN implementations tracked by Optech, C-Lightning also has
  support for signet (see [Newsletter #117][news117 cl4068]).

- [LND #5155][] adds a configuration option to randomly select which
  wallet UTXOs to spend in a transaction; this reduces UTXO
  fragmentation in the wallet over time.  By contrast, the default coin
  selection algorithm in LND spends higher value UTXOs before lower
  value UTXOs; this minimizes fees in the short term but may result in
  needing to pay higher fees in the future when all inputs near the size
  of a transaction, or larger, have already been spent.

- [BOLTs #672][] updates [BOLT2][] to allow nodes to negotiate a
  `option_shutdown_anysegwit` option which, if set, allows LN closing
  transactions to be able to pay any segwit script version, including
  script types that don't yet have consensus meaning on the network,
  such as addresses for [taproot][topic taproot].

- [BOLTs #872][] updates [BOLT3][]'s use of [BIP69][] to specify in more
  detail the sort order to use for commitment transaction inputs and

>   outputs.  One commentator points out that the use of BIP69 has so far

Reviewer 4: why "one commentator"? seems inconsistent because [there have already been many mentions of Corallo](https://www.google.com/search?q=site%3Abitcoinops.org+%22matt+corallo%22)

Author: @ysangkok We almost never[1] mention names in the *notable changes* section.  The reason is that notable changes to large projects usually require the close cooperation of many participants---designers, implementers, reviewers, testers, maintainers, etc---and I worry it would be unfair and bad for morale if only a subset of those people received recognition for the whole group's effort.  In this particular case, several people who put quite a lot of debugging effort into identifying this as the cause of accidental channel force closures, while others came up with this simple solution and then implemented and tested it, and it would seem really unfair to me if the only name we mentioned out of all that work was Corallo's for a quick comment he made re-stating his objections to BIP69 (which are well known to other LN maintainers).

[1] Off hand, I'm only aware of one exception to this policy, see the last two sentences of https://bitcoinops.org/en/newsletters/2019/03/26/#bitcoin-core-10973

[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Bitcoin Core #20833][] is the first PR in an effort to bring [package
  relay][topic package relay] to Bitcoin Core. This change allows the
  `testmempoolaccept` RPC to accept multiple transactions. Future PRs will
  enable submitting transaction packages to the mempool through the RPC as well as

>   specifications for communicating packages over the P2P network.

Author: Maybe s/through the RPC/through other RPCs/ and s/specifications for/implement/
